### ISSUE TITLE
Bad value: order_tax_amount

### ISSUE THREAD

**Issue by yratof** [krokedil/klarna-checkout-for-woocommerce#38]
Can't seem to get the checkout working:

`kco-iframe`

![image](https://user-images.githubusercontent.com/1001935/38608429-98529140-3d7b-11e8-918c-956b2182dfc5.png)

```
{\"type\":\"shipping_fee\",\"reference\":\"Shipping\",\"name\":\"Shipping\",\"quantity\":1,\"unit_price\":0,\"tax_rate\":0,\"total_amount\":0,\"total_tax_amount\":0}],\"shipping_countries\":[\"NO\"],\"billing_address\":{\"email\":null,\"postal_code\":null,\"country\":\"NO\",\"phone\":null,\"given_name\":null,\"family_name\":null,\"street_address\":null,\"street_address2\":null,\"city\":null,\"region\":null},\"options\":{\"allow_separate_shipping_address\":true,\"date_of_birth_mandatory\":true,\"national_identification_number_mandatory\":true,\"title_mandatory\":true,\"allowed_customer_types\":[\"person\"]},\"customer\":{\"type\":\"person\"}}"}
2018-04-11T11:40:27+00:00 ERROR stdClass::__set_state(array(
   'error_code' => 'BAD_VALUE',
   'error_messages' => 
  array (
    0 => 'Bad value: order_tax_amount',
  ),
   'correlation_id' => '542c01b4-4d05-4f5e-9a0d-f11f203cdca2',
   'service_version' => '1.0.20180405120507',
))
2018-04-11T11:40:29+00:00 ERROR Retrieve ongoing Klarna order (https://api.playground.klarna.com/checkout/v3/orders/) {"headers":{"Authorization":"Basic UEswMTkzMV80MDQ0ZWI2ZTE0MzA6TnFkZnNjTWcxVGpoekxyQw==","Content-Type":"application\/json"},"user-agent":"WordPress\/4.8.3; http:\/\/fjellr-2255.rask22.raskesider.no\/fjellrevenshop - KCO:1.3.0"}
2018-04-11T11:40:29+00:00 ERROR stdClass::__set_state(array(
   'error_code' => 'HTTP_ERROR',
   'error_messages' => 
  array (
    0 => '',
  ),
   'correlation_id' => '05da5aab-9a90-454f-b3e6-e0485477ed2e',
   'service_version' => '1.0.20180405120507',
))
```

**Comment by yratof** [comment]
Found this in relation: https://stackoverflow.com/questions/48050138/bad-value-order-tax-amount

**Comment by kristibarbosky** [comment]
@yratof - the request sent to Klarna for the merchant_create_order_v3 call matching your correlation_id shows that the "order_tax_amount" : 121900 doesn't add up correctly for the order line item tax amounts (39960 + 39980 + 41980), so Klarna validation throws the error (off by 20).  Do you have any plugins active in your store that are affecting tax calculations?  Perhaps can you deactivate other plugins in your store to see if the problem still exists?  If this doesn't help you determine the root cause of the incorrect tax calculations, could you provide your WooCommerce->Status->System Status Report so we can attempt to recreate this in a vanilla Woo store.  

Also, you can review the Klarna merchant portal to see your Logs to provide additional information, such as the request body of the merchant_create_order_v3 call.

**Comment by MichaelBengtsson** [comment]
@yratof In the WooCommerce settings, how many decimals do you have it set to display? We need this to be set to 2 decimals for calculations to work properly. If you still want prices displayed with 0 decimals you can use a code snippet that you can find here: [https://gist.github.com/NiklasHogefjord/72a79fc3a8b9e8f052d82a5acc4d53a9]
And put that in your child-themes functions.php file.

**Comment by yratof** [comment]
Currently sticking with the default 2:

![image](https://user-images.githubusercontent.com/1001935/38617228-b07be9ea-3d96-11e8-8bb1-b785686b8e23.png)

Looked at the plugins used currently & there's nothing that affects woocommerce — just ACF and SEO

**Comment by yratof** [comment]
Here's something strange — I set the decimals to 0 -still didn't work, but then went back to 2 again and it kicked in...

**Comment by yratof** [comment]
Ok the issue is if you set the decimal amount to 0, `Bad value: order_tax_amount` is shown in the iframe

**Comment by MichaelBengtsson** [comment]
@yratof Yes sadly that is something that we cant fix. This is due to how WooCommerce does the calculations of the tax, and its based on that setting.

**Comment by yratof** [comment]
So if we set the decimal point to 0, Klarna has to fail? Whats the work around for this?

**Comment by forsvunnet** [comment]
Hi @MichaelBengtsson. This issue was fixed two years ago, but was never merged. See: https://github.com/krokedil/woocommerce-gateway-klarna-private/pull/230

**Comment by MichaelBengtsson** [comment]
@yratof As i posted above, for merchants that still want to display zero decimals you can use the code found here: https://gist.github.com/NiklasHogefjord/72a79fc3a8b9e8f052d82a5acc4d53a9

@forsvunnet Thank you for that link, i will take a look at it and see if we can implement it. I did not know of that commit.

**Comment by MichaelBengtsson** [comment]
@forsvunnet We need to investigate this a bit more, as if i understand it correctly this could mean that we get a different price sent to klarna then is shown in WooCommerce?

**Comment by forsvunnet** [comment]
@MichaelBengtsson No, you'll be sending the correct price.
Let me give you an example:

```
// Item price is 12.50 inc 25% VAT
// VAT = 12.50 * 0.25 = 3.125

// With rounding (Wrong):
$item_tax_percentage = $item['line_tax'] * 100 / $order->get_line_total( $item, false );
// 3.125 * 100 / 13 = 24.04

// Without rounding (Correct):
$item_tax_percentage = $item['line_tax'] * 100 / $order->get_line_total( $item, false, false );
// 3.125 * 100 / 12.50 = 25
```

**Comment by forsvunnet** [comment]
@MichaelBengtsson I might be wrong. The code I'm refering to is the old klarna plugin, but the error message is the exact same as we had back then. I suspect it's the same issue, but the code looks to be entirely different.

**Comment by MichaelBengtsson** [comment]
@forsvunnet With the new plugin we are using the cart instead of order. So will have to see if this applies to that. But im looking into that now. Thank you for the code example and explination though!

**Comment by forsvunnet** [comment]
@MichaelBengtsson I investigated a bit further. It all comes down to how WooCommerce calculates totals in the cart. It might be worth raising an issue with them to change the WC core code.

Look at how totals are calculated in WC_Cart::calculate_totals
```
do_action( 'woocommerce_before_calculate_totals', $this );
new WC_Cart_Totals( $this );
do_action( 'woocommerce_after_calculate_totals', $this );
```

If you then follow the construct all the way through WC_Cart_Totals::calculate_totals you find that it uses WC_Cart_Totals::get_merged_taxes to calculate the tax total.
It again takes a parameter $in_cents which in the case of tax_total is false. The `$in_cents` is a toggle for whether the result is passed through `wc_remove_number_precision_deep` or not. That means that already here the tax total is rounded.

If that was not enough, back in the WC_Cart_Totals::calculate_totals the resulting tax total is assigned using the WC_Cart::set_total_tax method. This method is basically just a wrapper to write the value to the WC_Cart::$totals array using `wc_round_tax_total`.

TLDR; It looks to me that cart methods really aren't meant to do anything beyond displaying values to the front end. I recommend that you drop using WC_Cart and instead implement the code using WC_Order methods

**Comment by forsvunnet** [comment]
I took another deeper look this morning.
Alarmingly the "Number of decimals" also affect orders and order data. It does no longer only affect the display price. This is an issue with the WooCommerce core itself.

Here's a workaround I made to get the correct totals:
```
class Test
{
  /**
   * Get unrounded totals from WooCommerce
   * @return array
   */
  public function get_unrounded_totals() {
    // Save the already calculated totals
    $original_totals = WC()->cart->get_totals();
    // Override the price_decimals
    add_filter( 'wc_get_price_decimals', __CLASS__.'::price_decimals' );
    // Re-calculate the cart using increased precision
    WC()->cart->calculate_totals();
    // Store the new totals
    $totals = WC()->cart->get_totals();
    // Revert changes
    remove_filter( 'wc_get_price_decimals', __CLASS__.'::price_decimals' );
    WC()->cart->set_totals( $original_totals );
    return $totals;
  }

  /**
   * Filter for price decimals
   * @param  integer $decimals
   * @return integer
   */
  static function price_decimals( $decimals ) {
    // Allow no less than 4 decimals
    if ( $decimals > 4 ) {
      return $decimals;
    }
    return 4;
  }
}
```

**Comment by forsvunnet** [comment]
I created an issue on WooCommerce here: https://github.com/woocommerce/woocommerce/issues/19783

**Comment by MichaelBengtsson** [comment]
@forsvunnet Thank you for the explination and example code, you really did a lot of work on this. Sadly i see that the issue was closed, and i read the explination as to why. We will look at implementing the workarround that you provided, but will have to do some testing to se if it is viable or not with different situations. If you have any comments on why the ticket was closed by WooCommerce, i would love to hear it.

**Comment by forsvunnet** [comment]
@kristibarbosky WooCommerce's recommendation is that the sum total displayed on the checkout page should also be the sum paid in the transaction. Is that how payments should be made to Klarna?
The alternative is to use the actual unrounded price with decimals. What is the best approach for accounting etc?

**Comment by forsvunnet** [comment]
@MichaelBengtsson the alternative to recalculating the price with higher precision is to calculate the tax_total based on the tax_rate. Eg:

```php
// $tax_total = 88 - 88 / ( 1 + 0.25 );
$tax_total = $order_total - $order_total / ( 1 + $tax_rate );
```

**Comment by kristibarbosky** [comment]
@forsvunnet - if non-zero decimals are included in the price, non-zero decimals must be sent to Klarna.  If the customer is to be charged non-zero decimals, the customer should be displayed non-zero decimals so the customer knows the exact amount they are required to pay.

**Comment by forsvunnet** [comment]
@kristibarbosky Ok, so the consensus is that what the user sees is what the user pays?
@MichaelBengtsson I recommend implementing the solution above where you calculate the tax-amount based on the tax rate instead of relying on the tax_total from WC_Cart.

**Comment by NiklasHogefjord** [comment]
Thanks @forsvunnet, we will test this in detail. Will get back to you by the end of this week.

**Comment by NiklasHogefjord** [comment]
I want to clarify a couple things what we are trying to solve here.

1. We are trying to enable the possibility to use the 0 decimal setting in WooCommerce and still being able to finalize a purchase with Klarna Checkout. 

Up until now this hasn't been possible since the tax amount isn't always correct in WC cart when using 0 decimals. We are working on a fix for this.

1. If we get this fix working we still (in some scenarios) might run into the issue with order totals that differ between WooCommerce and Klarna. To solve this we are looking into a adding an order line (in Klarna) that compensate for this difference (in the cases this happens).

I think it's very important that we all agree on the fact that it isn't possible to have each order line exactly the same in both systems when using 0 decimals in WooCommerce. This is not possible since Klarna wants the product price to be sent over with _price including tax_, _tax rate_ and _tax amount_.

Please let me know if you think that I'm missing the point with the original question/issue.

**Comment by forsvunnet** [comment]
@NiklasHogefjord 

1. Yes
2. No, the order total will be the same

You only need to change the `tax_total` that is being sent to klarna.

Eg. If the order total is rounded to 88 then a 25% tax is 17.6. You need to calculate the `tax_total` based on the `tax_rate` instead of using the `cart->tax_total` which is rounded to 18.

**Comment by NiklasHogefjord** [comment]
@forsvunnet plus use the rounded price including tax as unit_price?

**Comment by forsvunnet** [comment]
@NiklasHogefjord I might be wrong, but I don't think you have to change anything relating to the unit price. The error message from the Klarna API is caused by the `line_total` times the `tax_rate` not being equal to the `tax_total`.

**Comment by NiklasHogefjord** [comment]
@forsvunnet To be able to reflect the same amount in both systems it does look like the rounded price including tax must be used. I have spent quite some time with this now and and at the moment I'm looking into shipping tax calculation.

I'll get back to you on Monday with a status report, or hopefully a working branch for you to test. :)

**Comment by NiklasHogefjord** [comment]
Okay. I've got to the point where I can manage products, shipping and discounts fairly good with 0 decimals, but we have a scenario that might cause the iframe not to render. This is because WooCommerce seem to mis calculate the order in some cases.

I triggered this behaviour by doing the following:

**1. Set prices to be displayed with 0 decimals.**
![image](https://user-images.githubusercontent.com/985946/39676569-4d445bf6-516d-11e8-8a28-9e004c4ff7f5.png)

**2. Set a product with the price of 20,12 kr.**
![image](https://user-images.githubusercontent.com/985946/39676567-38ad130e-516d-11e8-88a5-93fab813db25.png)

This will show as _20 kr_ in admin overview and in front-end.

![image](https://user-images.githubusercontent.com/985946/39676517-54d9848c-516c-11e8-9b5a-4e2c32bde20c.png)
![image](https://user-images.githubusercontent.com/985946/39676581-8817c772-516d-11e8-9819-3a380d5781fb.png)

**3. Add 6 pieces of this product to cart.**
![image](https://user-images.githubusercontent.com/985946/39676593-c3652ef0-516d-11e8-82ad-8a9758671cd2.png)

**4. KCO iframe fails to render**
Because WooCommerce calculates 6 x 20 = 121.
![image](https://user-images.githubusercontent.com/985946/39676606-ec6f6d56-516d-11e8-9d8a-ec415873bdbf.png)

If I complete this order with _Direct Bank Transfer_ as payment method instead, we can see how Woo sums up the order:
![image](https://user-images.githubusercontent.com/985946/39676621-4720e342-516e-11e8-8ec0-1938b582f5e1.png)

Even with a quantity of 5 there I see this behaviour happen (WooCommerce calculates 5 x 20 = 121), but in this case Woo still manage to get the order total right.
![image](https://user-images.githubusercontent.com/985946/39676680-1a444138-516f-11e8-8b71-04101964415d.png)

Perhaps this is just an edge case and a bug in WooCommerce. I'm not sure. If it is we can probably aim to add support for 0 decimals in the plugin. I will push the branch tomorrow so you can test.

**Comment by NiklasHogefjord** [comment]
Hi again,
Sorry for the delay.

I've now pushed the updates to this branch - https://github.com/krokedil/klarna-checkout-for-woocommerce/tree/develop-zero-decimal-fix.

It would be great if you could test it @forsvunnet and see if it works with your setup.

@kristibarbosky This also solves issue #39.

**Comment by forsvunnet** [comment]
@NiklasHogefjord It almost works! If you have prices inclusive tax then the calculation works, but if you have prices exclusive tax then it's broken.

![image](https://user-images.githubusercontent.com/799281/40004551-3a05f2d2-5796-11e8-8309-a83f35200f5e.png)

**Comment by NiklasHogefjord** [comment]
Thanks for the report @forsvunnet! I've updated the branch now. Can you give it a go again?

**Comment by jonasostrom** [comment]
Encountered the issue today when using a coupon with percentage value.
Decimals: 0
Coupon with 25% discount

Widget: Bad value: total_tax_amount

**Comment by NiklasHogefjord** [comment]
@jonasostrom Did you get this error message when using the fix in this branch: https://github.com/krokedil/klarna-checkout-for-woocommerce/tree/develop-zero-decimal-fix ?

**Comment by NiklasHogefjord** [comment]
@forsvunnet Did you get a chance to test the updated version of the branch?

**Comment by jonasostrom** [comment]
@NiklasHogefjord Not yet, Is it up to date with current master?

**Comment by NiklasHogefjord** [comment]
@jonasostrom not at the moment. I'll try to do that later today.

**Comment by yratof** [comment]
@forsvunnet has this been solved and tested as working? I'm still waiting for a fix, as all my prices look like:

![image](https://user-images.githubusercontent.com/1001935/41662384-4cbc4a66-74a1-11e8-9d7f-5f32e1dfb02e.png)

**Comment by yratof** [comment]
@NiklasHogefjord Just installed on a test site now, i can confirm that it's still not working.

`Bad value: total_tax_amount`

**Comment by NiklasHogefjord** [comment]
@yratof Latest version merged into https://github.com/krokedil/klarna-checkout-for-woocommerce/tree/develop-zero-decimal-fix now so it should be up to date.

Can you give me a description of the product price plus product quantity when you add it to cart to trigger the issue?

**Comment by yratof** [comment]
Just pulled the merge, it seems to be working fine in the local environment, will be pushing to a test site shortly to test further @NiklasHogefjord

**Comment by inthujan** [comment]
I'm also having this problem, would be great if you can release the update anytime soon so that I don't have to use decimals.

**Comment by MichaelBengtsson** [comment]
@inthujan Do you have the same issue as yratof or do you have decimals that are .00? If its the later case then you can use this filter in your themes function.php file to get arround the issue while still having 2 decimals in your store: https://gist.github.com/NiklasHogefjord/72a79fc3a8b9e8f052d82a5acc4d53a9

**Comment by inthujan** [comment]
@MichaelBengtsson I have tried this and it does work but my shop has a few products with decimals that are not .00 so it doesn't work for those and we wish to round up the price using no decimals :(

**Comment by Pautomagi** [comment]
Im facing a somewhat similar issue to whats explained here.
I managed to reproduce the tax error, but my main issue is that the new checkout wont render in the iframe. Didn't get an error either so i started investigating.

Dumping the get order from the `klarna-checkout-for-woocommerce-functions.php` files function `kco_wc_show_snippet()`

`var_dump(KCO_WC()->api->get_order());`

Gets me this:
```
object(WP_Error)#10128 (2) { ["errors"]=> array(1) { ["kco"]=> array(1) { [0]=> string(23) "Bad value: order_amount" } } ["error_data"]=> array(0) { } }
```

Dumping the variable `$klarna_order` only gives me this:
```
object(WP_Error)#10166 (2) { ["errors"]=> array(0) { } ["error_data"]=> array(0) { } }
```

**Comment by yratof** [comment]
@inthujan @MichaelBengtsson @Pautomagi Have you all tried the `develop-zero-decimal-fix` branch? It solved the problem for the decimals set to 0.

Also @Pautomagi, the iFrame doesn't load if there's something wrong. Often it's the amounts not adding up right, (hence this thread). Since your problem says "Bad value: order_amount", i would presume it's related to products having decimals in their pricing, say, 10.1, and woocommerce rounding it down to display. But when you add 10 to cart, the total is 110, not 100 and klarna sees two values wrong.

Try the branch

**Comment by Pautomagi** [comment]
@yratof Yes I get why the error occurs but even with decimals, without decimals and trying the decimal fix branch it keeps failing.

I will try a totally fresh install (the one I'm on is a new install but with a non standard theme, it might be overriding something somewhere).

**Comment by NiklasHogefjord** [comment]
When having these settings:

_Default customer location: No location by default_
![image](https://user-images.githubusercontent.com/985946/43270680-f98ae3d6-90f5-11e8-96fa-b8de52f88974.png)

_Prices entered with tax: Yes, I will enter prices inclusive of tax_
![image](https://user-images.githubusercontent.com/985946/43270799-3857fb76-90f6-11e8-82ac-a9c899dcca88.png)

And add a product that costs 18 SEK (or any other currency) including 25.00% tax.
![image](https://user-images.githubusercontent.com/985946/43271184-2ccc9e8c-90f7-11e8-97d5-103b44a06270.png)

If I add 2 pcs of this product, then there will be a mismatch:
![image](https://user-images.githubusercontent.com/985946/43271259-5948abb8-90f7-11e8-8dae-f6166b865db9.png)

This happens because Woocommerce rounds per order line, not by item. 18 including 25% tax is 14,4 excluding tax. 14,4 x 2 = 28,8. WooCommerce then rounds this to 29. But we send it to Klarna as 14 x 2 = 28.

### Solution
To get 0 decimals to work with KCO I think we need to specify a couple of WooCommerce settings as mandatory.

1. Default customer address needs to be set to _Geolocate_ or _Shop base address_ so that prices includes tax from the first request. 
![image](https://user-images.githubusercontent.com/985946/43271515-03e0305a-90f8-11e8-9b2e-2d3221075b20.png)

2. Prices should be entered including tax.
![image](https://user-images.githubusercontent.com/985946/43272149-91a3bb36-90f9-11e8-9d70-254506f31884.png)

3. Prices should be entered as integers in the product, not with decimals.


@kristibarbosky perhaps you can set your test store with these settings and try to break the checkout? :)

**Comment by damircalusic** [comment]
Hi @NiklasHogefjord,

We have this problem as speeking. The problem here is that we are integrating an accounting system that needs the prices in decimals.

So the setting **Prices entered with tax** needs to be **No, I will enter prices exclusive of tax**. We have entered the other settings according to your suggestion.

The problem for us **(using the new Klarna V3)** is that this problem seems random. Yesterday morning a client tried to shop and got the error **Bad value - total tax amount** but when trying in the afternoon the client got no error.

What can be done regarding this?

**Comment by NiklasHogefjord** [comment]
Hi @damircalusic, do you use the develop-zero-decimal-fix branch in the webshop? 

Perhaps you can try to see if you can trigger the issue by adding at least 5 or 6 items of one product to cart and see if that might be what's causing the problem.

**Comment by damircalusic** [comment]
Just to be clear you are referring to this https://github.com/krokedil/klarna-checkout-for-woocommerce/tree/develop-zero-decimal-fix 

We are using the latest version of the Klarna V3 plguin (1.6.1) just incase you need to know.

**Comment by NiklasHogefjord** [comment]
OK, if you have 0 decimals and version 1.6.1, the issue you are describing will happen from time to time. Try changing to 2 decimals instead.

Or did I misunderstand you now? :)

**Comment by damircalusic** [comment]
Hmm.... let me clarify. We need to have 2 decimals. Which we have now. The prices in backend are shown with decimals, on the frontend without. So the settings are correct for us.

So, the error total tax amount appeears randomly :)

**Comment by NiklasHogefjord** [comment]
How do you change it to not display decimals in frontend?

**Comment by damircalusic** [comment]
We use `add_filter('woocommerce_price_trim_zeros', '__return_true');` to remove the decimals on the frontend

**Comment by NiklasHogefjord** [comment]
Thanks Damir, then this is a new kind of issue and not related to this exact issue. If you find out how to reproduce the error then you can create a new issue here. If you want help with troubleshooting this, perhaps you can get in direct contact with me? Ping me on Messenger or something similar :)

**Comment by michaellwatson** [comment]
how to calculate the tax_rate when building your own cart

t = total
y = tax amount
x = tax rate

t-(t/1+x)=y

x = y/(t-y)

**Comment by forsvunnet** [comment]
@NiklasHogefjord Any chance of this branch being merged into master? Currently, we have to do some convoluted server-side git merging when we update the plugin for our clients in order to use this branch.

**Comment by NiklasHogefjord** [comment]
Yes @forsvunnet, if you're using it and it does work then I'm satisfied :).

I will merge this, but I need to tweak the get_shipping_tax_rate() function a bit to get it to work with different tax classes a bit better.

Aiming for a new version in mid January.

**Comment by NiklasHogefjord** [comment]
So, @yratof, @forsvunnet, @damircalusic, @Pautomagi, @michaellwatson, @inthujan & @jonasostrom - the 0 decimal fix is now published live in version 1.8.1. 

I will do a write-up about things to consider when using 0 decimals together with KCO. I'll post it in this thread and also add that information to the docs.

However - our suggested way is still to use 2 decimals :). _If you send the order data to an external ERP or accounting system it might still not work with 0 decimals because of rounding_.

**Comment by MichaelBengtsson** [comment]
Closed since this should be resolved now. And we have not heard back from anyone regarding issues with this.

**Comment by yemi** [comment]
Im still getting:

> A mismatch in order totals between WooCommerce and Klarna was detected. Please try again.

When trying to checkout using Klarna, I added the 2 decimals configuration just in case (up from 0) but still seeing the same error, any ideas?

**Comment by NiklasHogefjord** [comment]
@yemi Please add a support ticket at support@krokedil.com if you want more in detail help regarding this.
